rfcs/proposed/host_backend_histogram/README.md

```diff
@@ -0,0 +1,143 @@
+also compound our issue of race conditions, multiplying the number of concurrent lines of execution by the vector
+length. The advantage we get from vectorization of the increment operation or the lookup into the output histogram may
+not provide much benefit, especially when we account for the extra memory footprint required or synchronization
+required to overcome the race conditions which we add from the additional concurrent streams of execution. It may make
+sense to decline to add vectorized operations within histogram depending on the implementation used, and based on
+performance results.
+
+## Existing Patterns
+
+### count_if
+`histogram` is similar to `count_if` in that it conditionally increments a number of counters based upon the data in a
+sequence. `count_if` returns a scalar-typed value and doesn't provide any function to modify the variable being
+incremented. Using `count_if` without significant modification would require us to loop through the entire sequence for
+each output bin in the histogram. From a memory bandwidth perspective, this is untenable. Similarly, using a
+`histogram` pattern to implement `count_if` is unlikely to provide a well-performing result in the end, as contention
+should be far higher, and `reduce` is a very well-matched pattern performance-wise.
+
+### parallel_for
+`parallel_for` is an interesting pattern in that it is very generic and embarrassingly parallel. This is close to what
+we need for `histogram`. However, we cannot simply use it without any added infrastructure. If we were to just use
+`parallel_for` alone, there would be a race condition between threads when incrementing the values in the output
+histogram. We should be able to use `parallel_for` as a building block for our implementation, but it requires some way
+to synchronize and accumulate between threads.
+
+## Proposal
+I believe there are two competing options for `histogram`, which may both have utility in the final implementation
+depending on the use case.
+
+### Implementation One (Embarrassingly Parallel)
+This method uses temporary storage and a pair of embarrassingly parallel `parallel_for` loops to accomplish the
+`histogram`.
+1) Determine the number of threads that we will use, perhaps adding some method to do this generically based on the
+2) backend.
+3) Create temporary data for the number of threads minus one copy of the histogram output sequence. Thread zero can
+4) use the user-provided output data.
+5) Run a `parallel_for` pattern which performs a `histogram` on the input sequence where each thread accumulates into
+6) its own copy of the output sequence using the temporary storage to remove any race conditions.
+7) Run a second `parallel_for` over the `histogram` output sequence which accumulates all temporary copies of the
+8) histogram into the output histogram sequence. This step is also embarrassingly parallel.
+9) Deallocate temporary storage.
```

Comment:
I think the changes to the line widths interfered with the numbering here.

Reply:
fixed, thanks